src/utils.ts

```diff
@@ -92,20 +93,75 @@ export class McpServerError extends Error {
 
 export function instrumentedMCPServer<T extends BaseMCPServer>(MCPServer: new (ctx: Context) => T, config: ResolveConfigFn) {
     const Agent = class extends McpAgent<Env, any, Props> {
-        server = new MCPServer(this);
+        private _server: T | undefined;
+        private _env: Env;
+
+        // Lazy getter for server that creates it when first accessed
+        // 
+        // WHY THIS APPROACH:
+        // Originally we passed 'this' directly as Context: `server = new MCPServer(this)`
+        // This worked when Context was just { props: Props }, but broke when we added env.
+        // 
+        // PROBLEMS WITH ORIGINAL APPROACH:
+        // 1. McpAgent's 'env' property is protected, but Context expects public
+        // 2. TypeScript error: "Property 'env' is protected but public in Context"
+        // 
+        // WHY NOT CONSTRUCTOR CREATION:
+        // We tried creating server in constructor: `new MCPServer({ props: this.props, env })`
+        // But this.props is undefined during constructor - it gets set later by McpAgent._init()
+        // Runtime error: "Cannot read properties of undefined (reading 'instanceUrl')"
+        // 
+        // SOLUTION - LAZY INITIALIZATION:
+        // - Store env from constructor (available immediately)
+        // - Create server only when first accessed (after props are set by McpAgent lifecycle)
+        // - Combine both props and env into proper Context object
+        get server(): T {
+            if (!this._server) {
+                const context: Context = {
+                    props: this.props, // Available after McpAgent._init() sets it
+                    env: this._env     // Stored from constructor
+                };
+                this._server = new MCPServer(context);
+            }
+            return this._server;
+        }
```

Comment:
declare public env: Env;

And then move creation of the server inside the `init` method.